lib/auth/helpers_mfa.go

```diff
@@ -0,0 +1,328 @@
+/*
+Copyright 2017-2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth/mocku2f"
+	"github.com/gravitational/teleport/lib/auth/u2f"
+	wanlib "github.com/gravitational/teleport/lib/auth/webauthn"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/pquerna/otp"
+	"github.com/pquerna/otp/totp"
+)
+
+// TestDeviceType represents the various supported MFA types.
+// TODO(codingllama): Replace with proto.DeviceType
+type TestDeviceType int
```

Comment:
Is anything in this file ever used outside of tests? I don't love that we have a lot of test code throughout the codebase that is easy to accidentally import in production code. Should we stick a`_test.go` suffix on this file so that it can't be compiled into non-test packages?

Reply:
If I stick a _test suffix we won't be able to use it in other packages, that is why I went for the "Test" prefix in the public interface.

TestTLSServer (and its friends) live alongside it in the lib/auth package, so there is precedent. Something I did in other projects was pushing things like this to a `testonly` package and/or using a "//+build testonly" build tag - both we could apply here, but I think that is beyond this PR.

Reply:
Fair :+1: